mm2src/mm2_main/src/lp_swap/maker_swap_v2.rs

```diff
@@ -485,133 +652,178 @@ impl<MakerCoin: MmCoin, TakerCoin: MmCoin + SwapOpsV2> State for TakerPaymentRec
+impl<MakerCoin: MmCoin + CoinAssocTypes, TakerCoin: MmCoin + SwapOpsV2> State
+    for MakerPaymentSentFundingSpendGenerated<MakerCoin, TakerCoin>
+{
     type StateMachine = MakerSwapStateMachine<MakerCoin, TakerCoin>;
 
     async fn on_changed(self: Box<Self>, state_machine: &mut Self::StateMachine) -> StateResult<Self::StateMachine> {
         let maker_payment_info = MakerPaymentInfo {
             tx_bytes: self.maker_payment.tx_hex.0.clone(),
             next_step_instructions: None,
+            funding_preimage_sig: self.funding_spend_preimage.signature.to_bytes(),
+            funding_preimage_tx: self.funding_spend_preimage.preimage.to_bytes(),
         };
         let swap_msg = SwapMessage {
             inner: Some(swap_message::Inner::MakerPaymentInfo(maker_payment_info)),
         };
 
         debug!("Sending maker payment info message {:?}", swap_msg);
-        let _abort_handle = broadcast_swap_v2_msg_every(
+        let abort_handle = broadcast_swap_v2_msg_every(
             state_machine.ctx.clone(),
             state_machine.p2p_topic.clone(),
             swap_msg,
             600.,
             state_machine.p2p_keypair,
         );
+
+        let recv_fut = recv_swap_v2_msg(
+            state_machine.ctx.clone(),
+            |store| store.taker_payment.take(),
+            &state_machine.uuid,
+            NEGOTIATION_TIMEOUT_SEC,
+        );
+        let taker_payment_info = match recv_fut.await {
+            Ok(p) => p,
+            Err(e) => {
+                let next_state = MakerPaymentRefundRequired {
+                    maker_coin_start_block: self.maker_coin_start_block,
+                    taker_coin_start_block: self.taker_coin_start_block,
+                    negotiation_data: self.negotiation_data,
+                    maker_payment: self.maker_payment,
+                    reason: MakerPaymentRefundReason::DidNotGetTakerPayment(e),
+                };
+                return Self::change_state(next_state, state_machine).await;
+            },
+        };
+        drop(abort_handle);
+
+        let taker_payment = match state_machine.taker_coin.parse_tx(&taker_payment_info.tx_bytes) {
+            Ok(tx) => tx,
+            Err(e) => {
+                let next_state = MakerPaymentRefundRequired {
+                    maker_coin_start_block: self.maker_coin_start_block,
+                    taker_coin_start_block: self.taker_coin_start_block,
+                    negotiation_data: self.negotiation_data,
+                    maker_payment: self.maker_payment,
+                    reason: MakerPaymentRefundReason::FailedToParseTakerPayment(e.to_string()),
+                };
+                return Self::change_state(next_state, state_machine).await;
+            },
+        };
+
         let input = ConfirmPaymentInput {
-            payment_tx: self.taker_payment.tx_hex.0.clone(),
+            payment_tx: self.taker_funding.tx_hex(),
```

Comment:
Next state is `TakerPaymentConfirmed`, shouldn't this be the `taker_payment` not `taker_funding`. We will also require a confirmation for the `taker_funding` in a previous step/state.

Reply:
Yeah, it should be taker payment, fixed.